src/main/java/persistence/sql/entity/PersistenceContextImpl.java

```diff
@@ -7,35 +7,76 @@
+    private final Map<EntityKey, Object> managedEntities = new HashMap<>();
+    private final Map<EntityKey, Object> entitySnapshots = new HashMap<>();
 
     @Override
     public <T> T getEntity(Class<T> clazz, Long id) {
-        if (!containsEntity(clazz, id)) {
+        EntityKey entityKey = new EntityKey(id, clazz);
+        Object entity = managedEntities.get(entityKey);
+
+        if (!containsEntity(entityKey)) {
             return null;
         }
-        return clazz.cast(managedEntities.get(clazz).get(id));
+        return clazz.cast(entity);
     }
 
     @Override
     public void addEntity(Object entity, Long id) {
-        Class<?> clazz = entity.getClass();
-        Long idValue = getIdValue(entity);
+        EntityKey entityKey = new EntityKey(id, entity.getClass());
+        managedEntities.put(entityKey, entity);
 
-        Map<Long, Object> longObjectMap = managedEntities.computeIfAbsent(clazz, aClass -> new HashMap<>());
-        longObjectMap.put(idValue, entity);
+        addSnapshot(id, entity);
     }
 
     @Override
     public void removeEntity(Class<?> clazz, Long id) {
-        if (containsEntity(clazz, id)) {
-            managedEntities.get(clazz).remove(id);
-        }
+        EntityKey entityKey = new EntityKey(id, clazz);
+        managedEntities.remove(entityKey);
+        entitySnapshots.remove(entityKey);
+    }
+
+    @Override
+    public boolean containsEntity(EntityKey entityKey) {
+        return managedEntities.containsKey(entityKey);
+    }
+
+    @Override
+    public Object getDatabaseSnapshot(Long id, Object entity) {
+        EntityKey entityKey = new EntityKey(id, entity.getClass());
+        return entitySnapshots.get(entityKey);
+    }
+
+    @Override
+    public void addSnapshot(Long id, Object entity) {
+        EntityKey entityKey = new EntityKey(id, entity.getClass());
+
+        Object snapshot = copySnapshot(entity, id);
+        entitySnapshots.put(entityKey, snapshot);
     }
 
     @Override
-    public boolean containsEntity(Class<?> clazz, Long id) {
-        return managedEntities.containsKey(clazz) && managedEntities.get(clazz).containsKey(id);
+    public boolean isDirty(Long id, Object currentEntity) {
+        EntityKey entityKey = new EntityKey(id, currentEntity.getClass());
+        Object entitySnapshot = entitySnapshots.get(entityKey);
+        return !currentEntity.equals(entitySnapshot);
+    }
+
+    private Object copySnapshot(Object entity, Long id) {
+        Class<?> clazz = entity.getClass();
+        Object snapshot;
+        try {
+            snapshot = clazz.getDeclaredConstructor().newInstance();
+            for (Field field : clazz.getDeclaredFields()) {
+                field.setAccessible(true);
+                Object value = field.get(entity);
+                field.set(snapshot, value);
+            }
+        } catch (Exception e) {
+            throw new RuntimeException("스냅샷 생성 실패");
+        }
```

Comment:
사소한 팁이지만 예외를 재정의 하실 때는 항상 이전 예외를 체이닝해주시는 것을 권장드립니다!
예외가 체이닝되지 않는다면 오류가 난 근본적인 원인을 파악하기가 어렵기 때문에 꼭 함께 챙겨주시면 좋을 것 같아요 :)
```suggestion
        } catch (Exception e) {
            throw new RuntimeException("스냅샷 생성 실패", e);
        }
```

Reply:
체이닝을 하면 예외를 추적하기 편해서 디버깅할 때 유리할 것 같네요 👍